src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+#if SIZE_MAX == UINT64_MAX
+    /* Swap odd and even bits. */
+    v = ((v >> 1) & 0x5555555555555555) | ((v & 0x5555555555555555) << 1);
+    /* Swap consecutive pairs. */
+    v = ((v >> 2) & 0x3333333333333333) | ((v & 0x3333333333333333) << 2);
+    /* Swap nibbles. */
+    v = ((v >> 4) & 0x0F0F0F0F0F0F0F0F) | ((v & 0x0F0F0F0F0F0F0F0F) << 4);
+    /* Reverse bytes. */
+    v = __builtin_bswap64(v);
+#else
+    /* 32-bit version. */
+    v = ((v >> 1) & 0x55555555) | ((v & 0x55555555) << 1);
+    v = ((v >> 2) & 0x33333333) | ((v & 0x33333333) << 2);
+    v = ((v >> 4) & 0x0F0F0F0F) | ((v & 0x0F0F0F0F) << 4);
+    v = __builtin_bswap32(v);
+#endif
+    return v;
+}
+
+/* Advances a scan cursor to the next value. It increments the reverse bit
+ * representation of the masked bits of v. This algorithm was invented by Pieter
+ * Noordhuis. */
+size_t nextCursor(size_t v, size_t mask) {
+    v |= ~mask; /* Set the unmasked (high) bits. */
+    v = rev(v); /* Reverse. The unmasked bits are now the low bits. */
+    v++;        /* Increment the reversed cursor, flipping the unmasked bits to
+                 * 0 and increments the masked bits. */
+    v = rev(v); /* Reverse the bits back to normal. */
+    return v;
+}
+
+/* The reverse of nextCursor. */
+static size_t prevCursor(size_t v, size_t mask) {
+    v = rev(v);
+    v--;
+    v = rev(v);
+    v = v & mask;
+    return v;
+}
+
+/* Returns 1 if cursor A is less then cursor B, compared in cursor next/prev
+ * order, 0 otherwise. This function can be used to compare bucket indexes in
+ * probing order (since probing order is cursor order) and to check if a bucket
+ * has already been rehashed, since incremental rehashing is also performed in
+ * cursor order. */
+static inline int cursorIsLessThan(size_t a, size_t b) {
+    /* Since cursors are advanced in reversed-bits order, we can just reverse
+     * both numbers to compare them. If a cursor with more bits than the other,
+     * it is not significant, since the more significatnt bits become less
+     * significant when reversing. */
+    return rev(a) < rev(b);
+}
+
+/* Rehashes one bucket. */
+static void rehashStep(hashset *s) {
+    assert(hashsetIsRehashing(s));
+    size_t idx = s->rehashIdx;
+    bucket *b = &s->tables[0][idx];
+    int pos;
+    for (pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+        if (!(b->presence & (1 << pos))) continue; /* empty */
+        void *element = b->elements[pos];
+        uint8_t h2 = b->hashes[pos];
+        /* Insert into table 1. */
+        uint64_t hash;
+        /* When shrinking, it's possible to avoid computing the hash. We can
+         * just use idx has the hash, but only if we know that probing didn't
+         * push this element away from its primary bucket, so only if the
+         * bucket before the current one hasn't ever been full. */
+        if (s->bucketExp[1] < s->bucketExp[0] && !s->tables[0][prevCursor(idx, expToMask(s->bucketExp[0]))].everfull) {
```

Comment:
could be `<=` for comparing exponents. afaik we do an equal-size rehash when we need to clean up everfull markers